plugins/catalog-backend/src/ingestion/processors/GitlabReaderProcessor.test.ts

```diff
@@ -0,0 +1,263 @@
+      const withToken: ProviderConfig = {
+        target: '',
+        apiBaseUrl: '',
+        token: 'A',
+      };
+      const withoutToken: ProviderConfig = {
+        target: '',
+        apiBaseUrl: '',
+      };
+      expect(
+        (getApiRequestOptions(withToken).headers as any)['PRIVATE-TOKEN'],
+      ).toEqual('A');
+      expect(
+        (getApiRequestOptions(withoutToken).headers as any)['PRIVATE-TOKEN'],
+      ).toBeUndefined();
+    });
+  });
+
+  describe('getRawRequestOptions', () => {
+    it('inserts a token when needed', () => {
+      const withToken: ProviderConfig = {
+        target: '',
+        rawBaseUrl: '',
+        token: 'A',
+      };
+      const withoutToken: ProviderConfig = {
+        target: '',
+        rawBaseUrl: '',
+      };
+      expect(
+        (getRawRequestOptions(withToken).headers as any)['PRIVATE-TOKEN'],
+      ).toEqual('A');
+      expect(
+        (getRawRequestOptions(withoutToken).headers as any)['PRIVATE-TOKEN'],
+      ).toBeUndefined();
+    });
+  });
+
+  describe('getApiUrl', () => {
+    it('rejects targets that do not look like URLs', () => {
+      const config: ProviderConfig = { target: '', apiBaseUrl: '' };
+      expect(() => getApiUrl('a/b', config)).toThrow(/Incorrect URL: a\/b/);
+    });
+
+    it('happy path for gitlab', () => {
+      const config: ProviderConfig = {
+        target: 'https://gitlab.com',
+        apiBaseUrl: 'https://gitlab.com/api/v4',
+      };
+      expect(
+        getApiUrl(
+          'https://gitlab.com/a/b/blob/branchname/path/to/c.yaml',
+          config,
+        ),
+      ).toEqual(
+        new URL(
+          'https://gitlab.com/api/v4/projects/a/b/repository/files/path/to/c.yaml/raw?ref=branchname',
+        ),
+      );
+    });
+
+    it('happy path for self hosted', () => {
+      const config: ProviderConfig = {
+        target: 'https://git.mycompany.net',
+        apiBaseUrl: 'https://git.mycompany.net/api/v4',
+      };
+      expect(
+        getApiUrl(
+          'https://git.mycompany.net/a/b/blob/branchname/path/to/c.yaml',
+          config,
+        ),
+      ).toEqual(
+        new URL(
+          'https://git.mycompany.net/api/v4/projects/a/b/repository/files/path/to/c.yaml/raw?ref=branchname',
+        ),
+      );
+    });
+  });
+
+  describe('getRawUrl', () => {
+    it('rejects targets that do not look like URLs', () => {
+      const config: ProviderConfig = { target: '', apiBaseUrl: '' };
+      expect(() => getRawUrl('a/b', config)).toThrow(/Incorrect URL: a\/b/);
+    });
+
+    it('happy path for gitlab', () => {
+      const config: ProviderConfig = {
+        target: 'https://gitlab.com',
+        rawBaseUrl: 'https://gitlab.com',
+      };
+      expect(
+        getRawUrl(
+          'https://gitlab.com/a/b/blob/branchname/path/to/c.yaml',
+          config,
+        ),
+      ).toEqual(
+        new URL('https://gitlab.com/a/b/raw/branchname/path/to/c.yaml'),
```

Comment:
This is interesting to me - the old one had a different path for raw that was actually under the v4 api as well (`https://gitlab.example.com/api/v4/projects/12345/repository/files/my%2Fpath%2Fto%2Ffile.yaml/raw?ref=branch`) - so this new path works as well? Looks simpler - does it also support auth for private repos though?

Reply:
This does seem to break fetching files via the API, at least in my testing (both public and private). That has two be done with two API requests afaict since the api request for a file requires [a numeric project id.](https://docs.gitlab.com/ee/api/repository_files.html#get-raw-file-from-repository) My code had an extra request to [look up the project id](https://github.com/spotify/backstage/blob/master/plugins/catalog-backend/src/ingestion/processors/GitlabApiReaderProcessor.ts#L108)

I can try to help adapt that here if it would be helpful @therealadum since I have a private repo to test with

Reply:
@forrestwaters that would be great!

Reply:
@freben would we prefer reverting to the old method (v4 api raw request) or grabbing the project id and then making the api request?

Reply:
Probably safest to use the existing method (the one that works for @forrestwaters ) first and then propose the new method in a separate PR as an improvement suggestion (and probably to then ping him again). Does that make sense?